File: examples/java/src/main/java/org/wickra/timemachine/examples/Seek.java
package org.wickra.timemachine.examples;

import org.wickra.timemachine.TimeMachine;

/**
 * A runnable Java example: load a small recorded feed and reconstruct the market
 * snapshot at a past timestamp.
 *
 * <pre>
 *   mvn -q compile exec:java -Dexec.mainClass=org.wickra.timemachine.examples.Seek
 * </pre>
 *
 * Every language example loads the same feed and prints the same summary.
 */
public final class Seek {
    private Seek() {}

    private static final String FEED =
            "{\"ts\":10,\"symbol\":\"SYM\",\"feed\":{\"kind\":\"market\",\"type\":\"trade\","
                    + "\"symbol\":{\"base\":\"AAA\",\"quote\":\"USDT\"},\"price\":\"100\","
                    + "\"quantity\":\"1\",\"aggressor\":\"Buy\",\"timestamp\":10}}\n"
                    + "{\"ts\":20,\"symbol\":\"SYM\",\"feed\":{\"kind\":\"market\",\"type\":\"trade\","
                    + "\"symbol\":{\"base\":\"AAA\",\"quote\":\"USDT\"},\"price\":\"110\","
                    + "\"quantity\":\"2\",\"aggressor\":\"Sell\",\"timestamp\":20}}";

    /** Embed the feed as a JSON string literal inside the load command. */
    private static String loadCommand() {
        StringBuilder sb = new StringBuilder("{\"cmd\":\"load\",\"data\":\"");
        for (int i = 0; i < FEED.length(); i++) {
            char c = FEED.charAt(i);
            if (c == '"') {
                sb.append("\\\"");
            } else if (c == '\n') {
                sb.append("\\n");
            } else {
                sb.append(c);
            }
        }
        sb.append("\"}");
        return sb.toString();
    }

    public static void main(String[] args) {
        try (TimeMachine tm = new TimeMachine("{}")) {
            tm.command(loadCommand());
            String snapshot = tm.command("{\"cmd\":\"seek\",\"ts\":20}");
            System.out.printf("wickra-timemachine %s%n", TimeMachine.version());
            // The snapshot JSON is byte-identical to every other binding; parse it
            // with a real JSON library in production. Here we just confirm the seek
            // reconstructed the SYM last price of 110.
            System.out.printf("SYM last present: %b%n", snapshot.contains("\"last\":110"));
        }
    }
}

File: examples/c/seek.c
/* A runnable C example: load a small recorded feed and reconstruct the market
 * snapshot at a past timestamp through the wickra-timemachine C ABI. Every
 * language example loads the same feed and prints the same summary. */
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#include "wickra_timemachine.h"

/* Two trades on SYM; records separated by a newline. */
static const char *FEED =
    "{\"ts\":10,\"symbol\":\"SYM\",\"feed\":{\"kind\":\"market\",\"type\":\"trade\",\"symbol\":{\"base\":\"AAA\",\"quote\":\"USDT\"},\"price\":\"100\",\"quantity\":\"1\",\"aggressor\":\"Buy\",\"timestamp\":10}}\n"
    "{\"ts\":20,\"symbol\":\"SYM\",\"feed\":{\"kind\":\"market\",\"type\":\"trade\",\"symbol\":{\"base\":\"AAA\",\"quote\":\"USDT\"},\"price\":\"110\",\"quantity\":\"2\",\"aggressor\":\"Sell\",\"timestamp\":20}}";

/* Build {"cmd":"load","data":"<JSON-escaped feed>"} into out. */
static void build_load(char *out, size_t cap) {
    size_t n = 0;
    n += (size_t)snprintf(out + n, cap - n, "{\"cmd\":\"load\",\"data\":\"");
    for (const char *p = FEED; *p && n + 2 < cap; p++) {
        if (*p == '"') {
            out[n++] = '\\';
            out[n++] = '"';
        } else if (*p == '\n') {
            out[n++] = '\\';
            out[n++] = 'n';
        } else {
            out[n++] = *p;
        }
    }
    snprintf(out + n, cap - n, "\"}");
}

static char *run(WickraTimeMachine *tm, const char *cmd) {
    int len = wickra_timemachine_command(tm, cmd, NULL, 0);
    if (len < 0) {
        return NULL;
    }
    char *buf = (char *)malloc((size_t)len + 1);
    if (buf) {
        wickra_timemachine_command(tm, cmd, buf, (size_t)len + 1);
    }
    return buf;
}

int main(void) {
    WickraTimeMachine *tm = wickra_timemachine_new("{}");
    if (!tm) {
        fprintf(stderr, "failed to build time machine\n");
        return 1;
    }
    char load[2048];
    build_load(load, sizeof(load));

    char *loaded = run(tm, load);
    char *snapshot = run(tm, "{\"cmd\":\"seek\",\"ts\":20}");
    if (!loaded || !snapshot) {
        fprintf(stderr, "command failed\n");
        free(loaded);
        free(snapshot);
        wickra_timemachine_free(tm);
        return 1;
    }

    printf("wickra-timemachine %s\n", wickra_timemachine_version());
    printf("snapshot bytes: %zu\n", strlen(snapshot));

    free(loaded);
    free(snapshot);
    wickra_timemachine_free(tm);
    return 0;
}
